### PR TITLE
#1877 Improve decrypt performance

### LINKED ISSUE #1877: FlowCrypt ios app crash when trying to decrypt a very large message
Description:
I have recently tested the ios app regarding the security report we received and found out that the app crashed when trying to decrypt an encrypted message with 100Mb lengths of text. 

Such payload can be created through python: `python3 -c 'print("A"*100000000)' > 100_million_A.txt` and encrypt the plain text message using `gpg`

Steps to reproduce:

1. Create a file containing a large input. eg. python3 -c 'print("A"*100000000)' > 100_million_A.txt
2. Encrypt it with `gpg` and send it yourself using Gmail's built-in message composer.
3. Open the message on the FlowCrypt iOS and it will crash the app.

The android app has a limit of cutting 50Kb of decrypted messages to prevent this (though the android app is not working as expected yet since it cuts the encrypted content, not the decrypted content) - https://github.com/FlowCrypt/flowcrypt-android/blob/c5d8e4b841c784a0101f20a3854a84f406b89987/FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/fragment/MessageDetailsFragment.kt#L1744

Please see https://github.com/FlowCrypt/flowcrypt-security/issues/215 for more details.

### PR DESCRIPTION
**Pull request by sosnovsky**
This PR increases `Core` performance by replacing `JavaScriptCore` with `WKWebView` (it's faster and also runs in separate process from app, which greatly reduces app memory and cpu usage)

Before migration decrypting 100mb message crashed app (because it was using several GBs of memory), and now it takes 10-15 seconds.

For now I haven't found how to run native swift functions through `WKWebView` (for increasing decrypting speed), but it currently doesn't affect overall app performance because of better `WKWebView` performance (created task for it - https://github.com/FlowCrypt/flowcrypt-ios/issues/1909)

I also changed `webpack` config to use `production` mode, which reduced generated js file size from 4.4mb to 2.9mb.

close #1877

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities
- [x] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

File: Core/source/core/pgp-key.ts [hunk 1]
```diff
@@ -383,6 +380,16 @@ export class PgpKey {
       lastModified,
       revoked: k.revocationSignatures.length > 0
     };
+
+    if (k.isPrivate()) {
+      Object.assign(keyDetails, {
+        private: k.armor(),
+        isFullyDecrypted: isFullyDecrypted(k),
+        isFullyEncrypted: isFullyEncrypted(k)
+      });
```
Comment (ioanmo226) [inline Core/source/core/pgp-key.ts:389]: Maybe better to return spread object(I don't prefer object assign though)
https://eslint.org/docs/latest/rules/prefer-object-spread
```ts
return {
    ...keyDetails,
    {
        private: k.armor(),
        isFullyDecrypted: isFullyDecrypted(k),
        isFullyEncrypted: isFullyEncrypted(k)
  }
};

```

Comment (sosnovsky) [reply Core/source/core/pgp-key.ts:389]: I like spread operator too for it's simplicity, but here we'll need to have 2 `return` to make it work:
```ts
if (k.isPrivate()) {
    return {
        ...keyDetails, 
        {
             private: k.armor(), 
             isFullyDecrypted: isFullyDecrypted(k),
             isFullyEncrypted: isFullyEncrypted(k)
        } 
    }
} else {
    return keyDetails
}

```

Comment (ioanmo226) [reply Core/source/core/pgp-key.ts:389]: Yeah, we indeed need it (we don't need else though).
But it's not bad for code performance and readability i think?

Comment (ioanmo226) [reply Core/source/core/pgp-key.ts:389]: I'm fine with your approach though

**Review by ioanmo226 — DISMISSED** [review]
Changes look great 👍 
Both performance and code quality improved 👍

File: FlowCrypt/Core/Core.swift [hunk 2]
```diff
@@ -27,71 +27,94 @@ protocol KeyParser {
...
-    private init() {}
+    private init() {
+        setupWebView()
+    }
+
+    // MARK: - Setup
+    func setupWebView() {
+        guard let jsFileSrc = getCoreJsFile() else { return }
+
+        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
```
Comment (tomholub) [inline FlowCrypt/Core/Core.swift:63]: the `??` should be `[unknown version]` or similar

**Comment by ioanmo226** [comment]
@sosnovsky FYI I pushed small lint fixes to not disturb you.
By the way, did you try to login with latest changes? I got below error.

Next priority should be to fix `live test` I think. Because live test doesn't work now, semaphoreci won't notice us if something is broken in live.

image

**Review by tomholub — DISMISSED** [review]
Is the webview process reused for more than one call? Because it probably takes a few hundred ms to initialize. I suppose the answer is yes.

**Comment by sosnovsky** [comment]
> Is the webview process reused for more than one call? Because it probably takes a few hundred ms to initialize. I suppose the answer is yes.

@tomholub Yes, after initialization web view is reused for all calls. During testing I haven't noticed any issues with that - web view is initialized on `Core` init which happens before all calls.

**Comment by sosnovsky** [comment]
> @sosnovsky FYI I pushed small lint fixes to not disturb you. 

Thanks!

> By the way, did you try to login with latest changes? I got below error.

Yes, I logged in several times today - worked well for me, and I think login code wasn't changed in this PR.

> Next priority should be to fix `live test` I think. Because live test doesn't work now, semaphoreci won't notice us if something is broken in live.

Yeah, I tried to re-enable it, but with no luck - gmail still asks for additional confirmation.

File: FlowCrypt/Core/Core.swift (hunk 2 above)
Comment (sosnovsky) [reply FlowCrypt/Core/Core.swift]: Done ✅

**Comment by ioanmo226** [comment]
Hmm... not sure why it says `id token expired` in my device.
> Yes, I logged in several times today - worked well for me, and I think login code wasn't changed in this PR.

**Comment by sosnovsky** [comment]
@ioanmo226 I just re-checked google login - no issues with it, does it still fail for you?

**Comment by sosnovsky** [comment]
@ioanmo226 does it happen just after you tap `Continue with Gmail` or after you entered your login details? Did you try to login to another account?

**Comment by sosnovsky** [comment]
@ioanmo226 I found such issue in `AppAuth` repo - https://github.com/openid/AppAuth-iOS/issues/443, maybe your device time is wrong too?

**Comment by ioanmo226** [comment]
@sosnovsky You are right.
My device time was incorrect(I changed it because i had to test something)
I changed time to correct time (only had 2 hours difference though) and it's fine now.

Thank you.

**Review by ioanmo226 — APPROVED** [review]
👍

**Pull request merged**